File: lua/problem_36.lua
local function newBST(val, left, right)
    return { val = val, left = left, right = right }
end

local function secondLargest(root)
    local parent, curr = nil, root
    while curr.right do
        parent = curr; curr = curr.right
    end
    if curr.left then
        local node = curr.left
        while node.right do node = node.right end
        return node.val
    end
    return parent.val
end

local function problem36()
    print("Problem 36:")
    local root = newBST(5,
        newBST(3, newBST(2,nil,nil), newBST(4,nil,nil)),
        newBST(7, newBST(6,nil,nil), newBST(8,nil,nil))
    )
    assert(secondLargest(root) == 7, "assertion failed")
    local root2 = newBST(3, nil, newBST(4, nil, newBST(5,nil,nil)))
    assert(secondLargest(root2) == 4, "assertion failed")
    local root3 = newBST(5, newBST(3, nil, newBST(4,nil,nil)), nil)
    assert(secondLargest(root3) == 4, "assertion failed")
    print("All Tests Passed")
end

return { problem36 = problem36 }

File: typescript/problem_36.ts
class BSTNode36 {
    constructor(public val: number, public left: BSTNode36 | null = null, public right: BSTNode36 | null = null) {}
}

export function secondLargest(root: BSTNode36): number {
    let parent: BSTNode36 | null = null;
    let curr: BSTNode36 = root;
    while (curr.right) { parent = curr; curr = curr.right; }
    if (curr.left) {
        let node = curr.left;
        while (node.right) node = node.right;
        return node.val;
    }
    return parent!.val;
}

export function problem36(): void {
    console.log("Problem 36:");
    const root = new BSTNode36(5,
        new BSTNode36(3, new BSTNode36(2), new BSTNode36(4)),
        new BSTNode36(7, new BSTNode36(6), new BSTNode36(8))
    );
    if (secondLargest(root) !== 7) throw new Error("assertion failed");
    const root2 = new BSTNode36(3, null, new BSTNode36(4, null, new BSTNode36(5)));
    if (secondLargest(root2) !== 4) throw new Error("assertion failed");
    const root3 = new BSTNode36(5, new BSTNode36(3, null, new BSTNode36(4)), null);
    if (secondLargest(root3) !== 4) throw new Error("assertion failed");
    console.log("All Tests Passed");
}
